File: Tree/Trie/Trie.cpp
#include <iostream>
#include <unordered_map>

class TrieNode {
public:
    std::unordered_map<char, TrieNode*> children;
    bool isEndOfWord;

    TrieNode() : isEndOfWord(false) {}
};

class Trie {
private:
    TrieNode* root;

public:
    Trie() : root(new TrieNode()) {}

    void insert(const std::string& word) {
        TrieNode* node = root;
        for (char c : word) {
            if (node->children.find(c) == node->children.end()) {
                node->children[c] = new TrieNode();
            }
            node = node->children[c];
        }
        node->isEndOfWord = true;
    }

    bool search(const std::string& word) {
        TrieNode* node = searchNode(word);
        return node != nullptr && node->isEndOfWord;
    }

    bool startsWith(const std::string& prefix) {
        TrieNode* node = searchNode(prefix);
        return node != nullptr;
    }

private:
    TrieNode* searchNode(const std::string& word) {
        TrieNode* node = root;
        for (char c : word) {
            if (node->children.find(c) != node->children.end()) {
                node = node->children[c];
            } else {
                return nullptr;
            }
        }
        return node;
    }
};

int main() {
    Trie trie;

    trie.insert("apple");
    trie.insert("app");
    trie.insert("apricot");

    std::cout << std::boolalpha;
    std::cout << trie.search("apple") << std::endl;   // true
    std::cout << trie.search("app") << std::endl;     // true
    std::cout << trie.search("apricot") << std::endl; // true
    std::cout << trie.search("ap") << std::endl;      // false

    std::cout << trie.startsWith("ap") << std::endl;   // true
    std::cout << trie.startsWith("banana") << std::endl; // false

    return 0;
}

File: Tree/Trie/Trie.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.is_end_of_word = False


class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word):
        node = self.root
        for char in word:
            if char not in node.children:
                node.children[char] = TrieNode()
            node = node.children[char]
        node.is_end_of_word = True

    def search(self, word):
        node = self._search_node(word)
        return node is not None and node.is_end_of_word

    def starts_with(self, prefix):
        node = self._search_node(prefix)
        return node is not None

    def _search_node(self, word):
        node = self.root
        for char in word:
            if char in node.children:
                node = node.children[char]
            else:
                return None
        return node


# Пример использования
trie = Trie()
trie.insert("apple")
trie.insert("app")
trie.insert("apricot")

print(trie.search("apple"))  # True
print(trie.search("app"))  # True
print(trie.search("apricot"))  # True
print(trie.search("ap"))  # False

print(trie.starts_with("ap"))  # True
print(trie.starts_with("banana"))  # False
